### PR TITLE
IO-3-dev Importing command from text

### PR DESCRIPTION
**Pull request by BartoszKowalczyk98**

### CODE REVIEW

**Comment by BartoszKowalczyk98** [comment]
Przykładowa zawartość pliku Json:
{
  "commandName": "jakis eseflores",
  "operations": [
    {
      "opName": "SetPositionCommand",
      "x": -20,
      "y": -50
    },
    {
      "opName": "OperateToCommand",
      "x": -20,
      "y": -50
    },
    {
      "opName": "OperateToCommand",
      "x": 20,
      "y": 50
    },
    {
      "opName": "OperateToCommand",
      "x": 25,
      "y": 50
    }
  ]
}

File: src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java [hunk 1]
```diff
@@ -83,6 +102,25 @@ public CommandManagerWindow(DriverCommandManager commandManager) {
 		content.add(btnRunCommand, c);
 	}
 
+	private void importCommands() {
+		int returnValue = fileChooser.showOpenDialog(null);
+		if (returnValue == JFileChooser.APPROVE_OPTION) {
+			File selectedFile = fileChooser.getSelectedFile();
+			JsonParser jsonParser = new JsonParser(selectedFile);
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java]: The goal is to make the design of parser reusable. Now we accept the content only from a File (what if we have a simple String? we are forced to save it to File to use parser 👎 ) and we cannot add support for other formats easily.
Moreover, we cannot reuse parser instance (we need a new one for a single parse). This all should be fixed by defining a correct interface.

File: src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java [hunk 2]
```diff
@@ -83,6 +102,25 @@ public CommandManagerWindow(DriverCommandManager commandManager) {
...
+		}
+	}
+
+	private void exportCommands() {
+		fileChooser.setDialogTitle("Specify a file to save");
+		int userSelection = fileChooser.showSaveDialog(this);
+		if (userSelection == JFileChooser.APPROVE_OPTION) {
+			File fileToSave = fileChooser.getSelectedFile();
+			JsonParser jsonParser = new JsonParser(fileToSave);
+			jsonParser.parseToExport(commandManager.getCurrentCommand());
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java]: The same problem

File: src/main/java/edu/kis/powp/jobs2d/command/manager/parsing/JsonParser.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+import edu.kis.powp.jobs2d.command.OperateToCommand;
+import edu.kis.powp.jobs2d.command.SetPositionCommand;
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.List;
+
+public class JsonParser implements Parser {
+	File jsonFile;
+	String commandName = "No command has been set yet";
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/manager/parsing/JsonParser.java]: These should be private (however, I do not like a dependency to a File)

File: src/main/java/edu/kis/powp/jobs2d/command/manager/parsing/JsonParser.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+			JsonDriverCommand jsonDriverCommand = gson.fromJson(reader, JsonDriverCommand.class);
+			List<DriverCommand> driverCommands = new ArrayList<>();
+			for (JsonDriverCommand.operation op : jsonDriverCommand.operations)
+				if (op.opName.equals("SetPositionCommand")) driverCommands.add(new SetPositionCommand(op.x, op.y));
+				else if (op.opName.equals("OperateToCommand")) driverCommands.add(new OperateToCommand(op.x, op.y));
+			commandName = jsonDriverCommand.commandName;
+			return driverCommands;
+		} catch (FileNotFoundException e) {
+			e.printStackTrace();
+		}
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/manager/parsing/JsonParser.java]: weak exception handling

File: src/main/java/edu/kis/powp/jobs2d/command/manager/parsing/Parser.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+import java.io.File;
+import java.util.List;
+
+public interface Parser {
+	List<DriverCommand> parseFromImport();
+
+	String getCommandName();
+
+	File parseToExport(DriverCommand driverCommand);
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/manager/parsing/Parser.java]: Again, why File?

File: src/main/java/edu/kis/powp/jobs2d/command/manager/parsing/Parser.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package edu.kis.powp.jobs2d.command.manager.parsing;
+
+import edu.kis.powp.jobs2d.command.DriverCommand;
+
+import java.io.File;
+import java.util.List;
+
+public interface Parser {
+	List<DriverCommand> parseFromImport();
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/manager/parsing/Parser.java]: The interface is inconsistent!!! It defines an export format, but no import format. Internal format for import is List of commands and for export, it is a DriverCommand.

File: src/main/java/edu/kis/powp/jobs2d/drivers/macro/MacroDriverDecorator.java [hunk 7]
```diff
@@ -40,11 +40,13 @@ public Job2dDriver getDriver() {
 	@Override
 	public void setPosition(int x, int y) {
 		driverCommandList.add(new SetPositionCommand(x, y));
+		driver.operateTo(x,y);
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/drivers/macro/MacroDriverDecorator.java]: What happened here?! The lines added here are not related to your task? Where did you get those (copy&paste?)?!

**Review by kowallus — CHANGES_REQUESTED** [review]
Some issues, suggestions and questions?

File: src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java (hunk 1 above)
Comment (AndrzejBilant) [reply src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java]: At "we cannot add support for other formats easily". There is a filter method in FileChooser and all we have to do is add format after a comma.

File: src/main/java/edu/kis/powp/jobs2d/drivers/macro/MacroDriverDecorator.java (hunk 7 above)
Comment (BartoszKowalczyk98) [reply src/main/java/edu/kis/powp/jobs2d/drivers/macro/MacroDriverDecorator.java]: This was to solve that "Bug: drawing does not work during recording MACRO" and it slipped with one commit and i didn't notice it go now but it is now drawing lines during recording macro. So that's what happend.

Comment (kowallus) [reply src/main/java/edu/kis/powp/jobs2d/drivers/macro/MacroDriverDecorator.java]: OK, but it would be nice to have it in a separate PR (but this require working on different branches on different features)

File: src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java [hunk 8]
```diff
@@ -58,6 +69,43 @@ public CommandManagerWindow(DriverCommandManager commandManager) {
...
+			} catch (IOException exception) {
+				exception.printStackTrace();
+			}
+		});
+		c.fill = GridBagConstraints.BOTH;
+		c.weightx = 1;
+		c.gridx = 0;
+		c.weighty = 1;
+		content.add(btnExportCommand, c);
+
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java]: This method is large now. It would be best to decompose it to several smaller.

File: src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java [hunk 9]
```diff
@@ -83,6 +131,24 @@ public CommandManagerWindow(DriverCommandManager commandManager) {
 		content.add(btnRunCommand, c);
 	}
 
+	private void importCommands() {
+		parser = new JsonParser();
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java]: This parser could be a local variable (it is used this way). It would be better if the JsonParser object could be reused (then it can be a field).

File: src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java [hunk 10]
```diff
@@ -83,6 +131,24 @@ public CommandManagerWindow(DriverCommandManager commandManager) {
...
+		commandManager.setCurrentCommand(parser.parseFromString(txtFieldToImport.getText()));
+	}
+
+	private void exportCommands() throws IOException {
+		fileChooser.setDialogTitle("Specify a file to save");
+		int userSelection = fileChooser.showSaveDialog(this);
+		if (userSelection == JFileChooser.APPROVE_OPTION) {
+			File fileToSave = fileChooser.getSelectedFile();
+			BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(fileToSave));
+			parser = new JsonParser();
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java]: the same issue.

File: src/main/java/edu/kis/powp/jobs2d/command/manager/parsing/JsonParser.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+		Gson gson = new Gson();
+		JsonDriverCommand jsonDriverCommand = new JsonDriverCommand();
+
+		jsonDriverCommand.operations = getOperationFromDriverCommand(driverCommand);
+		return gson.toJson(jsonDriverCommand, JsonDriverCommand.class);
+	}
+
+	private List<JsonDriverCommand.operation> getOperationFromDriverCommand(DriverCommand driverCommand) {
+		List<JsonDriverCommand.operation> operationList = new ArrayList<>();
+		if (driverCommand instanceof OperateToCommand) {
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/manager/parsing/JsonParser.java:51]: Only an optional suggestion: instanceof is not object-oriented approach. If you want to make it right you can use, e.g. visitor pattern.

**Review by kowallus — CHANGES_REQUESTED** [review]
One mandatory code cleaning suggestion.

**Comment by kowallus** [comment]
You can work on taking advantage of the visitor pattern to fix unnecessary instanceof as your second task.
